---
title: "Imputation Statistics"
author: "Shea Andrews and Brian Fulton-Howard"
output: html_document
# params:
#    path: NA
#    outpath: NA
#    chrom: NA
#    cohort: NA
#    maf: NA #cutoff between common and rare
#    rsq: NA #Rsq to use with common variants
#    rsq2: NA #Rsq to use with rare variants (set to NA to use rsq as hard cutoff)
#    wrsq: NA
#    sampsize: NA #number of SNPs of each type to sample for figures
#    reading: "P" #P for parallel, L for looping (how to read files in)
#    rwd: NA
---
```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = snakemake@params[["rwd"]])

##---- Requried Pacakges ----##
suppressPackageStartupMessages(library(dplyr)) ## data wrangling
library(tidyr)
library(tibble)
library(ggplot2)    ## plotting
library(scales)     ## for using commas on ggplot scales
## displaing tables nicly in html
suppressPackageStartupMessages(library(kableExtra))
library(stringr)    ## replace substring for file output
suppressPackageStartupMessages(library(magrittr))
suppressPackageStartupMessages(library(purrr))
library(furrr)
library(httr)
library(readr)

no_rsq2 <- (is.na(snakemake@params[["rsq2"]]) ||
            snakemake@params[["rsq2"]] == "NA")

if (!no_rsq2) {
  rsqs <- c(snakemake@params[["rsq"]], snakemake@params[["rsq2"]])
} else {
  rsqs <- snakemake@params[["rsq"]]
}

ishtml <- knitr::opts_knit$get("rmarkdown.pandoc.to") == "html"

samp_pretty <- prettyNum(snakemake@params[["sampsize"]],
                         big.mark = ",", preserve.width = "none")

## Parse the chromosome list

chr_parse <- function(chr_arg) {
  chr_arg <- unlist(strsplit(chr_arg, ","))
  chrom <- c()
  for (i in chr_arg) {
    ii <- unlist(strsplit(i, ":"))
    chrom <- append(chrom, ii[1]:ii[length(ii)])
  }
  return(chrom)
}

##  Read in chromsome info files (chr*.info.gz) from HRC imputation, filtering
##  by the chromosome list.

choose_files <- function(path, chrom) {
  f_list <- list.files(path = path, pattern = "*.info.gz")
  files_use <- c()
  chroms_missing <- c()
  for (i in chrom) {
    ch <- paste0("chr", i, ".info.gz")
    chrom_present <- grepl(ch, f_list)
    if (any(chrom_present)) {
      files_use[length(files_use) + 1] <- ch
    } else {
      chroms_missing[length(chroms_missing) + 1] <- i
      warning(sprintf("Chromosome %i is missing from disk.", i))
    }
  }
  return(files_use)
}
```

```{r data, include=FALSE, warning=FALSE, message=FALSE}
##---- Data Wrangling ----##

files <- choose_files(snakemake@params[["path"]],
                      chr_parse(snakemake@params[["chrom"]]))

##  Read info files into R
read_info <- function(f) {
  message(sprintf("Loading %s...", f))
  #Read MAF, Rsq, and EmpRsq as numeric, replacing "-" with NA
  f %>%
    vroom::vroom(col_types = "cccnnnncnnncc", na = c("-")) %>%
    #Seperate CHR:POS columns
    separate(SNP, c("CHR", "POS"), sep = ":") %>%
    #Turn Position into an integer
    mutate(POS = as.integer(POS)) %>%
    #Make SNP column with CHROM:POS:REF:ALT
    unite(SNP, c(CHR, POS, "REF(0)", "ALT(1)"),
          sep = ":", remove = FALSE) %>%
    #Select only relevant columns
    select(SNP, CHR, POS, MAF, Rsq, EmpRsq, Genotyped)
}

path_files <- paste0(snakemake@params[["path"]], files)

snakemake@wildcards[["cohort"]] %>%
  paste0("_impStats.rda") %>%
  save.image()

if (snakemake@params[["reading"]] == "P") {
  plan(multicore(workers = 11))
} else {
  plan(sequential())
}

dat <- future_map_dfr(path_files, read_info)

# create new varible with MAF bins
mafbin_break <- c(-Inf, 0.0005, 0.001, 0.005, 0.01, 0.02, 0.05, Inf)
mafbin_labs <- c("< 0.05%", "0.05% to 0.1%", "0.1% to 0.5%",
                 "0.5% to 1%", "1% to 2%", "2% to 5%", "> 5%")
dat %<>% mutate(maf.cat = cut(MAF, breaks = mafbin_break,
                              labels = mafbin_labs))

# Relevel CHR
dat$CHR <- factor(str_replace(dat$CHR, "^chr", ""),
                  levels = c(1:22), labels = c(1:22))

## calculate number of typed/imputed SNPs
snps_imputed <- length(dat[dat$Genotyped == "Imputed", ]$SNP)
snps_typed <- length(dat[dat$Genotyped == "Genotyped", ]$SNP)

##  Subsample either the set number of imputed and genotyped SNPs
##    OR if less then the set number of SNPs avalibe, all SNPs
sampsize <- snakemake@params[["sampsize"]]
dat_smallsamp_imputed <- dat %>%
  filter(Genotyped == "Imputed") %>%
  group_by(CHR) %>%
  sample_frac(min(1, sampsize / snps_imputed), replace = FALSE) %>%
  # Don't use number greater than 1 if sample size is greater than the
  # actual number of SNPs.
  ungroup()

dat_smallsamp_typed <- dat %>%
  filter(Genotyped == "Genotyped") %>%
  group_by(CHR) %>%
  sample_frac(min(1, sampsize / snps_typed), replace = FALSE) %>%
  ungroup()

dat_smallsamp <- dat_smallsamp_imputed %>%
   bind_rows(dat_smallsamp_typed) %>%
   arrange(CHR, POS)
```

```{r data tabulation, echo=FALSE}
##---- Data Tabulation ----##
##  total number of imputed SNPs per chromosome
dat_out <- dat %>%
  mutate(Pass_R2_Common = (Rsq >= snakemake@params[["rsq"]] &
                           MAF >= snakemake@params[["maf"]])) %>%
  mutate(Pass_typed = Genotyped == "Genotyped")

if (no_rsq2) {
  dat_out %<>%
    mutate(Pass_MAF = MAF >= snakemake@params[["maf"]]) %>%
    mutate(included = (Pass_R2_Common & Pass_MAF) | Pass_typed)
} else {
  dat_out %<>%
    mutate(Pass_R2_Rare = (Rsq >= snakemake@params[["rsq2"]] &
                           MAF < snakemake@params[["maf"]]),
           included = Pass_R2_Rare | Pass_R2_Common | Pass_typed)
}

n_tested <- dat_out %>%
  select(Rsq, MAF, CHR) %>%
  filter_all(all_vars(!is.na(.))) %>%
  group_by(CHR) %>%
  group_size()

dat_out %<>%
  select(-EmpRsq, -maf.cat) %>%
  filter_all(all_vars(!is.na(.))) %>%
  select(CHR, matches("Pass_R2_Rare|Pass_MAF"),
         Pass_R2_Common, Pass_typed, included) %>%
  group_by(CHR) %>%
  summarise_all(sum) %>%
  ungroup() %>%
  tibble::add_column(n = n_tested, .after = 1) %>%
  mutate(CHR = as.character(CHR))

dat_out %<>%
  bind_rows(summarise_all(select(dat_out, -CHR), mean)) %>%
  bind_rows(summarise_all(select(dat_out, -CHR), sum))
dat_out[nrow(dat_out), "CHR"] <- "Total"
dat_out[nrow(dat_out) - 1, "CHR"] <- "Mean"

format_number <- function(x) {
  prettyNum(x, big.mark = ",", preserve.width = "none")
}

dat_out %<>%
  mutate(n = format_number(n),
         Pass_R2_Common = format_number(Pass_R2_Common),
         Pass_typed = format_number(Pass_typed),
         included = format_number(included))

if (no_rsq2) {
  dat_out %<>%
    rename(Pass_R2 = Pass_R2_Common) %>%
    mutate(Pass_MAF = format_number(Pass_MAF))
} else {
  dat_out %<>%
    mutate(Pass_R2_Rare = format_number(Pass_R2_Rare))
}

## save the table to disk
cohort_file <- str_replace(snakemake@params[["cohort"]], " ", "_")
out_snpcounts <-
  paste0(snakemake@params[["output_dir"]], "/SNPcount", cohort_file, ".csv")
write_csv(dat_out, out_snpcounts)
```

## Number of SNPs in datasets
The following table displays the number of imputed SNPs present in the GWAS, per chromosome and in total, pre- and post- SNP filtering for Minor allele frequency and Quality Score.

```{r kable, echo=FALSE}

dat_out %>%
  knitr::kable(format = "html", align = "lrrrr",
               caption = "Number of imputed SNPs, pre and post filtering") %>%
  kable_styling(full_width = FALSE, position = "center")

```
## Overall distribution of imputation quality

The following histogram is of imputation quality scores ($\hat{r}^2$) for all chromosomes and positions. The different colors represent different chromosomes.

```{r quality_distribution, warning=FALSE, echo = FALSE, fig.width=7.5, fig.height=4.7, fig.align="center"}
##  Plot histogram of info score distribution
ggplot(data = dat, aes(x = Rsq, fill = CHR)) +
  geom_histogram(binwidth = 0.01) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(labels = comma, expand=c(0, 0)) +
  geom_vline(xintercept = rsqs, colour = "red", linetype = 2) +
  theme_bw() +
  labs(title = paste0(snakemake@params[["cohort"]],
                      ": Distribution of imputation quality scores"),
       x = bquote("Quality ("~R^2~")"))
```

## Imputation quality in MAF ranges.

The following violin plots show the distribution of the imputation quality scores ($\hat{r}^2$) for all chromosomes and positions in several ranges of MAF from ultra-rare to common. This shows the minimum MAF you can use in your analyses and the general quality of the imputation, as well as any unusual distributions of quality.

```{r quality_mafbins, warning=FALSE, echo = FALSE, fig.width=7.5, fig.height=4.7, fig.align="center"}
##  Plot distribution of Rsq by MAF bins
ggplot(dat, aes(x = maf.cat, y = Rsq, fill = maf.cat)) +
  geom_violin(alpha = 0.75, scale = "width", draw_quantiles = 1) +
  geom_boxplot(width = 0.1, fill = "white", outlier.size = 0.1) +
  scale_fill_brewer(palette = "Set1") + theme(legend.position = "none") +
  theme_bw() +
  labs(title = paste0(snakemake@params[["cohort"]],
                      ": Distribution of Quality score by MAF "),
       x = "MAF",
       y = bquote("Quality ("~R^2~")"))
```

```{r kable maf_bins, echo=FALSE}

dat %>%
  group_by(maf.cat) %>%
  summarise(n = n(),
            "mean rsq" = mean(Rsq, na.rm = TRUE),
            "sd rsq" = sd(Rsq, na.rm = TRUE)) %>%
  rename("MAF Category" = maf.cat) %>%
  mutate(n = format_number(n)) %>%
  mutate_if(is.numeric, ~round(., 2)) %>%
  knitr::kable(format = "html", align = "lrcc",
    caption = "Number of imputed per MAF bin and mean rsq") %>%
  kable_styling(full_width = FALSE, position = "center")

```

## Position vs. Quality Score

The following scatter-plots show imputation quality scores ($\hat{r}^2$), plotted against position on each chromosome for aproxmitaly `r samp_pretty` imputed and `r samp_pretty` genotyped SNPs across the genome. This can be used to find gaps in imputation (there should only be gaps for the centromeres) or local dips in imputation qualtity.

```{r quality_pos, warning=FALSE, echo = FALSE, fig.width=7.5, fig.height=5.7, fig.align="center"}
## Plot Infoscore by genomic position
 ggplot(data = dat_smallsamp, aes(x = POS, y = Rsq)) +
  geom_point(size = 0.1, alpha = 0.5) +
  geom_hline(yintercept = rsqs, colour = "red", linetype = 1) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand=c(0, 0)) +
  facet_wrap(~CHR, scales = "free") +
  theme_classic() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  labs(title = paste0(snakemake@params[["cohort"]],
                      ": Quality score by genomic position"),
       x = "Genomic Position",
       y = bquote("Quality ("~R^2~")"))
```

## MAF vs. Quality Score

The following scatter-plot shows imputation quality scores ($\hat{r}^2$) plotted against logarithmically scaled Minor Allele Frequency (MAF). Because The relationship should be roughly exponential, it should appear sigmoid on the transformed plot. The place where the curve transitions will vary based on the successfulness of imputation.

The wave-like formations are due to imprecision in the MAF of the original data which is used to calculate $\hat{r}^2$. You can zoom to see that the peaks correspond to MAF bins in the typed data. The equation is as follows:

$$\hat{r}^2=\frac{\frac{1}{2n}\times\sum_{i=0}^{2n}(D_i-\hat{p})^2}{\hat{p}(1-\hat{p})}$$

where $\hat{p}$ is alternate allele frequency, $D_i$ is the imputed posterior probability of the alternate allele in the $i^{th}$ haplotype and $n$ is the number of microarray samples submitted to the imputation server.

Imputed SNPs are black, and typed SNPs are red. Approximately `r samp_pretty` imputed and `r samp_pretty` genotyped SNPs have been subsampled for the plot.

```{r quality_maf, warning=FALSE, echo = FALSE, fig.width=7.5, fig.height=4.7, fig.align="center"}
##  Plot info score vs MAF

plt <- dat_smallsamp %>%
  filter(Genotyped == "Genotyped") %>%
  rename(Quality = EmpRsq, Type = Genotyped) %>%
  select(MAF, Quality, SNP, Type)
plt <- dat_smallsamp %>%
  filter(Genotyped == "Imputed") %>%
  rename(Quality = Rsq, Type = Genotyped) %>%
  select(MAF, Quality, SNP, Type) %>%
  bind_rows(plt)

ggplot() +
  geom_point(data = plt,
             aes(x = MAF, y = Quality, color = Type),
             size = 0.1, alpha = 0.3) +
  scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x)),
    limits = c(0.00001, 0.5), expand = c(0, 0)) +
  scale_y_continuous(sec.axis = sec_axis(~ . +0,
                     name = bquote("Concordance (Emperical "~R^2~")")),
                     limits = c(0, 1), expand = c(0, 0)) +
  geom_vline(xintercept = snakemake@params[["maf"]],
             colour = "blue", linetype = 2) +
  geom_hline(yintercept = rsqs, colour = "red", linetype = 2) +
  theme_bw() +
  labs(title = paste0(snakemake@params[["cohort"]], ": Quality score by MAF"),
       y = bquote("Quality ("~R^2~")")) +
  scale_colour_manual(name = "SNPs", guide = "legend",
                      values = c("Imputed" = "black", "Genotyped" = "red")) +
  guides(colour = guide_legend(override.aes = list(size = 1))) +
  theme(legend.position = "bottom") +
  annotation_logticks()
```


```{R conditional_block_py, echo=FALSE, results="asis", eval=FALSE}
cat(sprintf("## MAF vs. Quality Score (interactive)

Below is an interactive version of the same plot, with approximately %s Genotyped SNPs and %s Imputed SNPs subsampled. You can zoom in and out, and you can mouse over individual variants for more information.

", samp_pretty, samp_pretty))
```

```{r ggploty, eval=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
library(plotly)
fig <- plt %>% plot_ly(x = ~MAF, y = ~Quality, color = ~Type,
  type = "scatter", mode = "markers",
  marker = list(size = 3, opacity = 0.3),
  colors = c("red", "black"), hoverinfo = "text+name",
  text = ~sprintf("MAF: %f\n%s: %f\nSNP: %s",
      MAF, if_else(Type == "Genotyped", "Empirical R²", "R²"), Quality, SNP),
  height = 600) %>%
  layout(title = paste0(snakemake@params[["cohort"]], ": Quality score by MAF"),
         xaxis = list(type = "log", range = c(-5, -0.3),
                      title = "Minor Allele Frequency (MAF)"),
         yaxis = list(range = c(0, 1),
            title = "Quality (R² for Imputed; Empirical R² for Typed)"),
         legend = list(orientation = "h")) %>%
  config(sendData = FALSE, displaylogo = FALSE,
         modeBarButtonsToRemove = list("sendDataToCloud", "lasso2d",
           "toImage", "autoScale2d")) %>%
  toWebGL()

shiny::div(
  style = "height:600;width:100%;border:none;margin:auto;display:block;", fig)
```
